tensorflow_decision_forests/contrib/temporal_feature_processor/temporal_feature_processor/implementation/pandas/operators/window/simple_moving_average.py

```diff
@@ -14,32 +14,72 @@
 
 """Implementation for the SimpleMovingAverage operator."""
 
+from typing import Dict
+
+import pandas as pd
+
 from temporal_feature_processor.implementation.pandas.data.event import PandasEvent
 from temporal_feature_processor.implementation.pandas.data.sampling import PandasSampling
 from temporal_feature_processor.implementation.pandas.operators.window.base import PandasWindowOperator
 
 
 class PandasSimpleMovingAverageOperator(PandasWindowOperator):
-  """Base class for window operators."""
+  """Pandas implementation for the simple moving average operator."""
 
-  def __init__(self, window_length: int) -> None:
+  def __init__(self, window_length: str) -> None:
     super().__init__(window_length=window_length)
 
   def __call__(
       self,
       data: PandasEvent,
       sampling: PandasSampling,
-  ) -> PandasEvent:
+  ) -> Dict[str, PandasEvent]:
     """Apply a simple moving average to an event.
 
     If input has more than one feature, the moving average will be computed for
     each of its features independently.
 
     Args:
-        data (PandasEvent): the input event to apply a simple moving average to.
+      data: the input event to apply a simple moving average to.
+      sampling: the desired sampling for the output event.
 
     Returns:
-        PandasEvent: the output of the operator.
+      Dict[str, PandasEvent]: the output event of the operator.
     """
+    # remove index to be able to filter using index values
+    data_no_index = data.reset_index()
+
+    # get index columns and name of timestamp column
+    index_columns = sampling.names[:-1]
+    timestamp_column = sampling.names[-1]
+
+    # create output event with desired sampling
+    output = PandasEvent(
+        {f"sma_{col}": [None] * len(sampling) for col in data.columns},
+        dtype=float).set_index(sampling)
+
+    # manual rolling window since pandas doesn't support custom sampling in .rolling()
+    # TODO: optimize window calculation
+    for values in sampling:
+      index_value = values[:-1]
+      timestamp = values[-1]
+
+      # filter by index_value
+      data_filtered = data_no_index[(
+          data_no_index[index_columns]
+          == index_value).squeeze()] if index_columns else data_no_index
+
+      # filter by window start/end dates
+      data_filtered = data_filtered[
+          (data_filtered[timestamp_column] <= timestamp) &
+          (data_filtered[timestamp_column] >= timestamp -
+           pd.Timedelta(self.window_length))]
+
+      # calculate average of window
+      mean = data_filtered.set_index(sampling.names).mean(axis=0).values
```

Comment:
Some questions:
- Will the mean fail or create an NaN or None if there are not values in the window?
- Can this output always be consumed  by a subsequent sma operator?
- Is the output a uniform type or a collection of object (of possibly different types)? I think it is slow for pandas to process columns contains different types of values (noted as "Object"). It would be interesting to research the pro/con of different solutions and then to define how all the operations should work. That another good topic for brainstorming :)